### ISSUE TITLE
Enhancement: [only-throw-error] support yielded and awaited errors

### ISSUE THREAD

**Issue by 43081j** [typescript-eslint/typescript-eslint#11418]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/only-throw-error

### Description

We can actually check yield and await expressions in the type checker it seems, so no longer need to bail when we encounter them.

### Fail

```TypeScript
async function foo() {
  throw await Promise.resolve<number>(303); // Not an Error
}
```

### Pass

```TypeScript
async function foo() {
  throw await Promise.resolve(new Error('error'));
}
```

### Additional Info

Draft here: #11417

**Comment by bradzacher** [comment]
I honestly have no idea why this early exit was added. Probably something to do with it being a port of a tslint rule? Either way no reason we should skip them I think.
